### PR TITLE
OpenCL support for cracking SSH keys - Initial Cut

### PR DESCRIPTION
**Pull request by kholia**
Note: This is still WIP.

On `i7-7820HQ` CPU,

```
$ ../run/john --test --device=2 --format=ssh 
Will run 8 OpenMP threads
Benchmarking: SSH [RSA/DSA/EC/OPENSSH (SSH private keys) 32/64]... (8xOMP) DONE
Speed for cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) of 1, cost 2 (iteration count) of 2
Raw:	3711K c/s real, 492867 c/s virtual
```

On GPU,

```
$ ../run/john --test --format=ssh-opencl
Device 6: GeForce GTX TITAN X
Benchmarking: ssh-opencl, SSH [MD5-3DES/AES OpenCL]... DONE
Speed for cost 1 (iteration count) of 2
Raw:	19660K c/s real, 16547K c/s virtual, GPU util: 89%
```

I was hoping for a better speedup. I wonder where the bottlenecks are. Is it the `3DES` stuff or the `ASN1` stuff?

### CODE REVIEW

File: src/opencl/ssh_kernel.cl [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+#ifndef PLAINTEXT_LENGTH
+#error PLAINTEXT_LENGTH must be defined
+#endif
+
+// input
+typedef struct {
+	uint length;
+        uchar v[PLAINTEXT_LENGTH];
+} ssh_password;
```
Comment (magnumripper) [inline src/opencl/ssh_kernel.cl]: looks like some kind of indentation problem here

Comment (kholia) [reply src/opencl/ssh_kernel.cl]: Yeah, the code is in very rough state at the moment. I will be fixing it soon.

**Comment by kholia** [comment]
@magnumripper Hey, is there a way I can I improve the auto-tuning for this format?

```
$ ../run/john --test=4 --device=6 --format=ssh-opencl
Device 6: GeForce GTX TITAN X
Benchmarking: ssh-opencl, SSH [MD5-3DES/AES OpenCL]... DONE
Speed for cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) of 0 and 1, cost 2 (iteration count) of 1 and 2
Raw:	18554K c/s real, 15625K c/s virtual, GPU util: 77%
```

```
$ ../run/john -lws=64 -gws=131072 --test=4 --device=6 --format=ssh-opencl
Device 6: GeForce GTX TITAN X
Benchmarking: ssh-opencl, SSH [MD5-3DES/AES OpenCL]... DONE
Speed for cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) of 0 and 1, cost 2 (iteration count) of 1 and 2
Raw:	20592K c/s real, 17096K c/s virtual, GPU util: 87%
```
```
$ ../run/john -lws=32 -gws=200000 --test=4 --device=6 --format=ssh-opencl
Device 6: GeForce GTX TITAN X
Benchmarking: ssh-opencl, SSH [MD5-3DES/AES OpenCL]... DONE
Speed for cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) of 0 and 1, cost 2 (iteration count) of 1 and 2
Raw:	21050K c/s real, 17432K c/s virtual, GPU util: 88%
```

**Comment by kholia** [comment]
The format works well on AMD devices I think but doesn't scale well on NVIDIA devices for some reason. Is there an OpenCL kernel profiler which I can use from Linux / macOS?

**Comment by kholia** [comment]
I am merging this now but I think that this will need more work (profiling and optimizations).

**Comment by kholia** [comment]
This fixes https://github.com/magnumripper/JohnTheRipper/issues/351 (implement ssh private key cracking on the GPU).

**Comment by magnumripper** [comment]
> @magnumripper Hey, is there a way I can I improve the auto-tuning for this format?

The auto-tune is probably fine. Look at the output of auto-tune with `-v:5`. You probably just need to optimize the kernel.

**Comment by magnumripper** [comment]
Oh, wait. You probably need GPU-side mask mode for this format!

**Comment by kholia** [comment]
> Oh, wait. You probably need GPU-side mask mode for this format!

The GPU utilization is already decent without a GPU-side mask mode. How will GPU-side mask mode help?

> The auto-tune is probably fine.

I posted some benchmarks -> with auto-tune and with hand-tweaked `lws+gws` values earlier in this thread. There is a good difference between the two I think.

**Comment by magnumripper** [comment]
> The GPU utilization is already decent without a GPU-side mask mode. How will GPU-side mask mode help?

No it isn't. 98% is decent.

```
$ ../run/john -form:ssh-opencl -test -v:5
Device 6: GeForce GTX TITAN X
Benchmarking: ssh-opencl [RSA/DSA/EC (SSH private keys) OpenCL]... Loaded 10 hashes with 10 different salts to test db from test vectors
(...)
Calculating best GWS for LWS=32; max. 200ms single kernel invocation.
Raw speed figures including buffer transfers:
xfer: 20.928us, crypt: 94.848us
gws:      3072	 26533Kc/s    26533996 rounds/s 115.776us per crypt_all()!
xfer: 46.016us, crypt: 104.960us
gws:      6144	 40695Kc/s    40695209 rounds/s 150.976us per crypt_all()+
xfer: 75.072us, crypt: 207.584us
gws:     12288	 43473Kc/s    43473338 rounds/s 282.656us per crypt_all()+
xfer: 147.232us, crypt: 407.808us
gws:     24576	 44277Kc/s    44277889 rounds/s 555.040us per crypt_all()+
(...)

Local worksize (LWS) 32, global worksize (GWS) 24576
DONE
Speed for cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) of 0 and 1, cost 2 (iteration count) of 1 and 2
Raw:	18751K c/s real, 16026K c/s virtual, GPU util: 78%
```
You spend 147us for transfer, 407us for crypt and maybe another 150us for result transfer which you somehow fail to measure (now THAT is an auto-tune problem BTW!). With GPU-side mask, perhaps you could spend over 700us for crypt... do the math.

**Pull request merged**